### DISCUSSION TITLE
v8.0.0 Roadmap and ETA

### DISCUSSION THREAD

**Discussion by PSanetra** [dotnet/dotnet-operator-sdk#629] [Q&A] [answered]
Hi, it seems there are a lot of breaking changes to be expected with the release of v8.0.0.

Is there some roadmap or ETA when the release of v8.0.0 might be expected? Would you recommand to start a new project for production use with v7 or do you expect already enough reliability with the current v8 prerelease?

(Of course additional api breaking changes are still to be expected when starting to use the v8 prerelease now.)

**Comment by buehler** [comment, marked as answer]
Hey @PSanetra 

Yes, there will be a lot of breaking changes, however, most of them should be fairly easy to migrate.
Feature wise, there are no breaks but the general "way of the sdk" did change a lot.

For example: there are more packages than just "KubeOps".

There is no fixed ETA, but at the time of writing this comment, I'm near the point of feature-equality to v7. Let's see how much time I can use in the near future ;-)

For new projects I suggest - *if you don't need webhooks* - to use the v8 since it will be the new way of the SDK. There are no future API breaks planned for controllers and events.

What is left for v8:
- Mutation webhooks
- Conversion webhooks
- Automatic installation of webhooks

What is left for v8, but is not important for the v8 release:
- Generator package should warn during build (custom roslyn analyzer) when webhooks are used without the proper attribute
- some way of developing helper for webhooks (like the localtunnel thingy, but that does not work anymore)

**Reply by PSanetra** [reply]
Thank you for the quick answer! I think in that case we will try to start with v8 :slightly_smiling_face: 

Regarding webhooks: You did not list `validating admission webhooks` for v8. Are they already supported with the current state of v8 or is it not planned to support them?

**Reply by buehler** [reply]
Right now, Validation Webhooks work :-)
you can develop a validation webhook but not the others.
I'm currently working on the fancy-installer-certificate generation stuff.

**Reply by buehler** [reply]
https://buehler.github.io/dotnet-operator-sdk/src/KubeOps.Operator.Web/README.html

**Reply by buehler** [reply]
@PSanetra automatic webhook installing should work as well. when you reference the operator.web package, it should (™️ ) generate the correct certificates and stuff for webhook operators.

**Reply by PSanetra** [reply]
Thank you for the information :+1: 
We currently plan to create two operators, working with the same CRDs. One should reconcile some k8s state and indeed also provide the validating webhook. The other should just reconcile the state of a subset of the CRD resources in the cluster. I guess in that case one operator should reference the `KubeOps.Operator.Web` package and the other one does not need to reference that package.

**Reply by buehler** [reply]
yes :-)
that was the plan (or part of it) of the big refactoring. you do not need to run asp.net if no webhooks are required. for any basic operator, the generic host is more than sufficient.

**Reply by ian-buse** [reply]
Do you have any updates on when v8 is expected? I pushed for my organization to allow me to contribute to this project, but we are on the v7 version and cannot migrate to v8 until it has reached parity with v7. The changes in v8 look great, and I'm looking forward to trying it out!

**Reply by buehler** [reply]
hey @ian-buse 
As far as I know, feature parity is achieved. The conversion webhooks are missing, but they were not present in v7.

**Comment by slacki123** [comment]
Hey @buehler, now that we're in mid January, how is the the release of v8 looking? Is there a predicted potential release date?

**Reply by buehler** [reply]
Hey @slacki123 
I've been slacking it 🙈 (btw: no pun intended)

Since I now have some feedback on the current features and the conversion webhooks "hidden" behind the preview attribute, we're kind of good to go. I want to make sure the newest PR is merged, and then I think I can release a V8. Hopefully I can fix the reconnection error as found by others first.

**Reply by slacki123** [reply]
Great news. Looking forward to it. Keep up the good work :)

**Reply by buehler** [reply]
... now... commit, push, run? 😂
